src/programs/bamseqchksum.1
.TH BAMSEQCHKSUM 1 "March 2014" BIOBAMBAM
.SH NAME
bamseqchksum - produce checksums for primary data in BAM files
.SH SYNOPSIS
.PP
.B bamseqchksum
[options]
.SH DESCRIPTION
bamseqchksum reads a BAM file from stdin, for each record calculates hash digest
checksums over
.IP [1]
flags and sequence
.IP [2]
queryname, flags and sequence
.IP [3]
flags, sequence and qualities
.IP [4]
flags, sequence and source data related aux tags
.PP
where the flags are the least significant byte of the BAM FLAGS containing
only the bits for multiple segments, first segment and last segment. The
sequence is reverse complemented, and quality string reversed, before 
checksumming if the reverse complemented bit is set.
.PP
Depending on the chosen hash digest function either the sum modulo some
power of 2 or the product modulo a prime number of these checksums is taken 
over all non-supplementary and non-secondary BAM alignment records. Separate 
sums or products are reported for combinations of all and QC pass records and for each
readgroup.
.LP
The following key=value pairs can be given:
.PP
.B verbose=<0>:
Valid values are
.IP 1:
print progress report on standard error
.IP 0:
do not print progress report
.PP
.B inputformat=<bam>: input file format
All versions of bamseqchksum come with support for the BAM input format. If
the program in addition is linked to the io_lib package, then the following
options are valid:
.IP bam:
BAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP sam:
SAM (see http://samtools.sourceforge.net/SAM1.pdf)
.IP cram:
CRAM (see http://www.ebi.ac.uk/ena/about/cram_toolkit)
.PP
.B reference=:
file name of the reference for CRAM input files. If this key is unset, then
the CRAM file header will be scanned for obtaining a reference file name.
.PP
.B hash=<crc32prod>:
hash digest used to compute checksums. All versions of biobambam support the
following functions:
.IP crc32prod:
checksums are computed via crc32 and combined over multiple records by multiplication modulo the prime number 2^31-1.
This is the default and only option for biobambam versions up to 0.0.174.
.IP crc32:
checksums are computed via crc32 and combined by summing up modulo 2^32.
.IP md5:
checksums are computed via md5 and combined by summing up modulo 2^128.
.IP crc32prime32:
identical with crc32prod (alternate implementation for testing purposes)
.IP crc32prime64:
checksums are computed via crc32 and combined over multiple records by multiplication modulo the prime number 2^64-59.
.IP md5prime64:
checksums are computed via md5 and combined over multiple records by multiplication modulo the prime number 2^64-59.
.IP crc32prime96:
checksums are computed via crc32 and combined over multiple records by multiplication modulo the prime number 2^96-17.
.IP md5prime96:
checksums are computed via md5 and combined over multiple records by multiplication modulo the prime number 2^96-17.
.IP crc32prime128:
checksums are computed via crc32 and combined over multiple records by multiplication modulo the prime number 2^128-159.
.IP md5prime128:
checksums are computed via md5 and combined over multiple records by multiplication modulo the prime number 2^128-159.
.IP crc32prime160:
checksums are computed via crc32 and combined over multiple records by multiplication modulo the prime number 2^160-47.
.IP md5prime160:
checksums are computed via md5 and combined over multiple records by multiplication modulo the prime number 2^160-47.
.IP crc32prime192:
checksums are computed via crc32 and combined over multiple records by multiplication modulo the prime number 2^192-237.
.IP md5prime192:
checksums are computed via md5 and combined over multiple records by multiplication modulo the prime number 2^192-237.
.IP crc32prime224:
checksums are computed via crc32 and combined over multiple records by multiplication modulo the prime number 2^224-63.
.IP md5prime224:
checksums are computed via md5 and combined over multiple records by multiplication modulo the prime number 2^224-63.
.IP crc32prime256:
checksums are computed via crc32 and combined over multiple records by multiplication modulo the prime number 2^256-189.
.IP md5prime256:
checksums are computed via md5 and combined over multiple records by multiplication modulo the prime number 2^256-189.
.IP null:
no checksums are computed and all checksums in the programs output are 0. This option is for performance testing only.
.PP
If libmaus is compiled with support for the nettle library, then the
following options are available:
.IP sha1:
checksums are computed via sha1 and combined by summing up modulo 2^160.
.IP sha1prime64:
checksums are computed via sha1 and combined over multiple records by multiplication modulo the prime number 2^64-59.
.IP sha1prime96:
checksums are computed via sha1 and combined over multiple records by multiplication modulo the prime number 2^96-17.
.IP sha1prime128:
checksums are computed via sha1 and combined over multiple records by multiplication modulo the prime number 2^128-159.
.IP sha1prime160:
checksums are computed via sha1 and combined over multiple records by multiplication modulo the prime number 2^160-47.
.IP sha1prime192:
checksums are computed via sha1 and combined over multiple records by multiplication modulo the prime number 2^192-237.
.IP sha1prime224:
checksums are computed via sha1 and combined over multiple records by multiplication modulo the prime number 2^224-63.
.IP sha1prime256:
checksums are computed via sha1 and combined over multiple records by multiplication modulo the prime number 2^256-189.
.IP sha224:
checksums are computed via sha2-224 and combined by summing up modulo 2^224.
.IP sha224prime64:
checksums are computed via sha2-224 and combined over multiple records by multiplication modulo the prime number 2^64-59.
.IP sha224prime96:
checksums are computed via sha2-224 and combined over multiple records by multiplication modulo the prime number 2^96-17.
.IP sha224prime128:
checksums are computed via sha2-224 and combined over multiple records by multiplication modulo the prime number 2^128-159.
.IP sha224prime160:
checksums are computed via sha2-224 and combined over multiple records by multiplication modulo the prime number 2^160-47.
.IP sha224prime192:
checksums are computed via sha2-224 and combined over multiple records by multiplication modulo the prime number 2^192-237.
.IP sha224prime224:
checksums are computed via sha2-224 and combined over multiple records by multiplication modulo the prime number 2^224-63.
.IP sha224prime256:
checksums are computed via sha2-224 and combined over multiple records by multiplication modulo the prime number 2^256-189.
.IP sha256:
checksums are computed via sha2-256 and combined by summing up modulo 2^256.
.IP sha256prime64:
checksums are computed via sha2-256 and combined over multiple records by multiplication modulo the prime number 2^64-59.
.IP sha256prime96:
checksums are computed via sha2-256 and combined over multiple records by multiplication modulo the prime number 2^96-17.
.IP sha256prime128:
checksums are computed via sha2-256 and combined over multiple records by multiplication modulo the prime number 2^128-159.
.IP sha256prime160:
checksums are computed via sha2-256 and combined over multiple records by multiplication modulo the prime number 2^160-47.
.IP sha256prime192:
checksums are computed via sha2-256 and combined over multiple records by multiplication modulo the prime number 2^192-237.
.IP sha256prime224:
checksums are computed via sha2-256 and combined over multiple records by multiplication modulo the prime number 2^224-63.
.IP sha256prime256:
checksums are computed via sha2-256 and combined over multiple records by multiplication modulo the prime number 2^256-189.
.IP sha384:
checksums are computed via sha2-384 and combined by summing up modulo 2^384.
.IP sha384prime64:
checksums are computed via sha2-384 and combined over multiple records by multiplication modulo the prime number 2^64-59.
.IP sha384prime96:
checksums are computed via sha2-384 and combined over multiple records by multiplication modulo the prime number 2^96-17.
.IP sha384prime128:
checksums are computed via sha2-384 and combined over multiple records by multiplication modulo the prime number 2^128-159.
.IP sha384prime160:
checksums are computed via sha2-384 and combined over multiple records by multiplication modulo the prime number 2^160-47.
.IP sha384prime192:
checksums are computed via sha2-384 and combined over multiple records by multiplication modulo the prime number 2^192-237.
.IP sha384prime224:
checksums are computed via sha2-384 and combined over multiple records by multiplication modulo the prime number 2^224-63.
.IP sha384prime256:
checksums are computed via sha2-384 and combined over multiple records by multiplication modulo the prime number 2^256-189.
.IP sha512:
checksums are computed via sha2-512 and combined by summing up modulo 2^512.
.IP sha512prime64:
checksums are computed via sha2-512 and combined over multiple records by multiplication modulo the prime number 2^64-59.
.IP sha512prime96:
checksums are computed via sha2-512 and combined over multiple records by multiplication modulo the prime number 2^96-17.
.IP sha512prime128:
checksums are computed via sha2-512 and combined over multiple records by multiplication modulo the prime number 2^128-159.
.IP sha512prime160:
checksums are computed via sha2-512 and combined over multiple records by multiplication modulo the prime number 2^160-47.
.IP sha512prime192:
checksums are computed via sha2-512 and combined over multiple records by multiplication modulo the prime number 2^192-237.
.IP sha512prime224:
checksums are computed via sha2-512 and combined over multiple records by multiplication modulo the prime number 2^224-63.
.IP sha512prime256:
checksums are computed via sha2-512 and combined over multiple records by multiplication modulo the prime number 2^256-189.
.IP sha512primesums:
checksums are computed via sha2-512 and combined over multiple records by adding modulo the Mersenne prime number 2^521-1.
.IP sha512primesums512:

debian/changelog
biobambam2 (2.0.30) unstable; urgency=medium

  * do not exit/fail bam indexing when encountering reads with mapping coordinates but flagged as unmapped as sole reads assigned to a reference sequence

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 17 Feb 2016 10:05:51 +0100

biobambam2 (2.0.29) unstable; urgency=medium

  * add murmur3 based digest option in bamseqchksum 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 11 Feb 2016 16:10:33 +0100

biobambam2 (2.0.28) unstable; urgency=medium

  * add bamrecalculatecigar program 

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 10 Feb 2016 14:14:04 +0100

biobambam2 (2.0.27) unstable; urgency=medium

  * add fastaexplode program 

 -- German Tischler <gt1@sanger.ac.uk>  Wed, 20 Jan 2016 15:27:26 +0100

biobambam2 (2.0.26) unstable; urgency=medium

  * update to libmaus2 2.0.114
  * Set variable to identify programs to iRODS. (by Andrew Whitwham)

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 12 Jan 2016 19:53:21 +0100

biobambam2 (2.0.25) unstable; urgency=medium

  * update to libmaus2 2.0.102 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 05 Nov 2015 15:28:29 +0100

biobambam2 (2.0.24) unstable; urgency=medium

  * update to libmaus2 2.0.99 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 18:33:56 +0100

biobambam2 (2.0.23) unstable; urgency=medium

  * update to libmaus2 2.0.98 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 02 Nov 2015 10:49:14 +0100

biobambam2 (2.0.22) unstable; urgency=medium

  * implicitly concatenate multiple input files as input for bamsort 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 29 Oct 2015 09:59:48 +0100

biobambam2 (2.0.21) unstable; urgency=medium

  * fix FF and RR type alignments in streaming duplicate marking
  * add MC tag (mate cigar string) when duplicate marking support is enabled in bamsort

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 26 Oct 2015 17:12:04 +0100

biobambam2 (2.0.20) unstable; urgency=medium

  * fix marking of FF and RR type duplicate records in BAM files 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 23 Oct 2015 14:55:46 +0200

biobambam2 (2.0.19) unstable; urgency=medium

  * include updates by Andrew Whitwham for improving the cigar handling of ClipAdapters

 -- German Tischler <gt1@sanger.ac.uk>  Tue, 22 Sep 2015 16:51:25 +0200

biobambam2 (2.0.18) unstable; urgency=medium

  * update to libmaus2 2.0.68 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 14 Sep 2015 14:03:00 +0200

biobambam2 (2.0.17) unstable; urgency=medium

  * allow building without requirement for SSE4 (change in libmaus2) 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 03 Sep 2015 10:39:07 +0200

biobambam2 (2.0.16) unstable; urgency=medium

  * merge Andrew Whitwham's patch for removing adapter clipping tags from secondary and supplementary alignments 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 24 Aug 2015 13:15:02 +0200

biobambam2 (2.0.15) unstable; urgency=medium

  * update to libmaus2 2.0.45 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 21 Aug 2015 10:13:18 +0200

biobambam2 (2.0.14) unstable; urgency=medium

  * update to libmaus2 2.0.42 

 -- German Tischler <gt1@sanger.ac.uk>  Sun, 09 Aug 2015 13:36:32 +0200

biobambam2 (2.0.13) unstable; urgency=medium

  * allow in memory tmp files in bammarkduplicates, bammarkduplicates2 and bamtofastq 

 -- German Tischler <gt1@sanger.ac.uk>  Sat, 08 Aug 2015 15:03:19 +0200

biobambam2 (2.0.12) unstable; urgency=medium

  * update to libmaus2 2.0.40

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 07 Aug 2015 08:36:53 +0200

biobambam2 (2.0.11) unstable; urgency=medium

  * update to libmaus2 2.0.39 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 06 Aug 2015 18:23:03 +0200

biobambam2 (2.0.10) unstable; urgency=medium

  * include Andrew Whitwham's (streaming) duplicate marking tag changes

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 22 Jun 2015 11:22:34 +0200

biobambam2 (2.0.9) unstable; urgency=medium

  * add daligner las to SAM/BAM/CRAM conversion tool 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 19 Jun 2015 15:45:53 +0200

biobambam2 (2.0.8) unstable; urgency=low

  * fix for random failures in bamsormadup due to missing/incomplete initialisation 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 27 Apr 2015 23:49:56 +0100

biobambam2 (2.0.7) unstable; urgency=low

  * update libmaus version to 2.0.6 (fix for bam indexing) 

 -- German Tischler <gt1@sanger.ac.uk>  Mon, 27 Apr 2015 11:43:45 +0100

biobambam2 (2.0.6) unstable; urgency=low

  * update libmaus version to 2.0.5 

 -- German Tischler <gt1@sanger.ac.uk>  Fri, 24 Apr 2015 09:31:27 +0100

biobambam2 (2.0.5) unstable; urgency=low

  * fix/clarify manual page for bamsormadup 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 23 Apr 2015 15:33:14 +0100

biobambam2 (2.0.4) unstable; urgency=low

  * add option to store temp data in RAM instead of on disc in bamsormadup 

 -- German Tischler <gt1@sanger.ac.uk>  Thu, 23 Apr 2015 13:53:51 +0100

biobambam2 (2.0.3) unstable; urgency=medium

  * update to libmaus 2.0.3 


src/programs/bamseqchksum.cpp
/**
    bambam
    Copyright (C) 2014-2014 David K. Jackson
    Copyright (C) 2009-2016 German Tischler
    Copyright (C) 2011-2014 Genome Research Limited

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
**/
#include <config.h>

#include <iostream>

#include <libmaus2/bambam/ChecksumsFactory.hpp>
#include <libmaus2/bambam/BamMultiAlignmentDecoderFactory.hpp>
#include <libmaus2/util/ArgInfo.hpp>

#include <biobambam2/Licensing.hpp>
#include <biobambam2/BamBamConfig.hpp>

#if defined(BIOBAMBAM_HAVE_GMP)
#include <libmaus2/math/UnsignedInteger.hpp>
#include <gmp.h>
#endif

static int getDefaultVerbose() { return 0; }
static std::string getDefaultInputFormat() { return "bam"; }
static std::string getDefaultHash() { return "crc32prod"; }
static int getDefaultStats() { return 0; }
static int getDefaultCheckHamming() { return 0; }

struct DComparatorBase
{
	uint64_t const d;
	std::vector<uint8_t> const & VV;

	DComparatorBase(uint64_t const rd, std::vector<uint8_t> const & rVV) : d(rd), VV(rVV) {}

	uint64_t hamming(uint64_t const i, uint64_t const j) const
	{
		uint64_t p0 = i*d;
		uint64_t p1 = j*d;
		uint64_t h = 0;

		for ( uint64_t k = 0; k < d; ++k )
		{
			uint64_t v0 = VV[p0+k];
			uint64_t v1 = VV[p1+k];
			uint64_t v2 = v0^v1;
			h += libmaus2::rank::PopCnt8<sizeof(unsigned long)>::popcnt8(v2);
		}
		return h;
	}

	uint64_t getBits(uint64_t const i, uint64_t from, uint64_t len) const
	{
		assert ( len <= 64 );

		unsigned int const rlen = len;

		uint64_t p0 = i*d;
		uint64_t const byteskip = from / 8;

		if ( byteskip )
		{
			p0 += byteskip;
			from -= byteskip * 8;
		}

		assert ( from < 8 );

		uint64_t w = 0;
		while ( len )
		{
			uint64_t v = VV[p0++];
			// bits available in this word/byte
			unsigned int const wlen = std::min(static_cast<uint64_t>(8-from),len);

			// align on left
			v <<= from;
			v &= 0xFF;
			// align on right
			v >>= (8-wlen);

			w <<= wlen;
			w |= v;

			from = 0;
			len -= wlen;
		}

		assert ( (rlen == 64) || w < (1ull<<rlen) );

		return w;
	}

	void print(std::ostream & out, uint64_t const i) const
	{
		uint64_t p0 = i*d;
		for ( uint64_t k = 0; k < d; ++k )
			out << std::hex << std::setw(2) << std::setfill('0') << static_cast<unsigned int>(VV[p0+k]) << std::setw(0) << std::dec;
	}
};

struct DComparator : public DComparatorBase
{
	DComparator(uint64_t const rd, std::vector<uint8_t> const & rVV) : DComparatorBase(rd,rVV) {}

	bool compare(uint64_t r0, uint64_t r1) const
	{
		uint64_t const p0 = r0*d;
		uint64_t const p1 = r1*d;

		for ( uint64_t k = 0; k < d; ++k )
		{
			uint8_t const v0 = VV[p0+k];
			uint8_t const v1 = VV[p1+k];

			if ( v0 < v1 )
				return true;
			else if ( v1 < v0 )
				return false;
		}

		return false;
	}

	uint64_t find(uint64_t rref, std::vector<uint64_t> const & I) const
	{
		uint64_t low = 0, high = I.size();

		while ( high-low > 1 )
		{
			uint64_t mid = (high+low)/2;

			if ( compare(rref,I[mid]) ) // rref < I[mid]
				high = mid;
			else // rref >= I[mid]
				low = mid;
		}

		return low;
	}

	bool operator()(uint64_t const i, uint64_t const j) const
	{
		return compare(i,j);
	}

	bool eq(uint64_t const i, uint64_t const j) const
	{
		return (!(*this)(i,j)) && (!(*this)(j,i));
	}
};

struct DBitComparator : public DComparatorBase
{
	unsigned int const from;
	unsigned int const len;

	DBitComparator(uint64_t const rd, std::vector<uint8_t> const & rVV, unsigned int const rfrom, unsigned int const rlen) : DComparatorBase(rd,rVV), from(rfrom), len(rlen) {}

	uint64_t get(uint64_t const i) const
	{
		return getBits(i,from,len);
	}

	bool operator()(uint64_t const i, uint64_t const j) const
	{
		uint64_t const v0 = get(i);
		uint64_t const v1 = get(j);
		return v0 < v1;
	}
};

void evaluateStats(std::vector<uint8_t> const & TV, std::vector<uint8_t> const & VV, bool checkHamming)
{
	uint64_t const d = TV.size();
	bool collisionfound = false;

	if ( d )
	{
		assert ( VV.size() % d == 0 );
		uint64_t const n = VV.size() / d;

		// bit counters
		std::vector<uint64_t> CC0(8*d,0);

		for ( uint64_t i = 0; i < n; ++i )
		{
			uint64_t o0 = i*d;
			uint64_t o = 0;

			for ( uint64_t j = 0; j < d; ++j )
			{
				uint8_t const v0 = VV.at(o0++);

				for ( unsigned int k = 0; k < 8; ++k )
				{
					uint64_t const b = (v0 >> (8-1-k)) & 1;
					CC0.at(o++) += b;
				}
			}
		}

		double maxabsbad = 0.0;
		double avgabsbad = 0.0;
		uint64_t absused = 0;
		for ( uint64_t i = 0; i < CC0.size(); ++i )
			if ( CC0[i] )
			{
				uint64_t const v = CC0.at(i);
				double const dv = v;
				double const frac = dv / n;
				double const bad = std::abs(0.5-frac);
				// std::cerr << "[S] abs msb " << i << " bad " << bad << std::endl;
				maxabsbad = std::max(bad,maxabsbad);
				avgabsbad += bad;
				absused += 1;
			}
		avgabsbad = absused ? avgabsbad/absused : avgabsbad;
		std::cerr << "[S] abs max bad " << maxabsbad << " avg bad " << avgabsbad << std::endl;

		// bit difference counters
		std::vector<uint64_t> CC(8*d,0);

		for ( uint64_t i = 1; i < n; ++i )
		{
			uint64_t o0 = (i-1)*d;
			uint64_t o1 = (i*d);
			uint64_t o = 0;

			for ( uint64_t j = 0; j < d; ++j )
			{
				uint8_t const v0 = VV.at(o0++);
				uint8_t const v1 = VV.at(o1++);
				uint8_t const v2 = v0 ^ v1;

				for ( unsigned int k = 0; k < 8; ++k )
				{
					uint64_t const b = (v2 >> (8-1-k)) & 1;
					CC.at(o++) += b;
				}
			}
		}

		if ( n > 1 )
		{
			double maxbad = 0;
			double avgbad = 0;

			uint64_t used = 0;
			for ( uint64_t i = 0; i < CC.size(); ++i )
				if ( CC[i] )
				{
					double const bad = std::abs(0.5 - static_cast<double>(CC[i]) / (n-1));
					maxbad = std::max(maxbad,bad);
					avgbad += bad;
					used += 1;
					//std::cerr << "MSB " << i << " prob " << static_cast<double>(CC[i]) / (n-1) << std::endl;
				}
			std::cerr << "[S] dif max bad " << maxbad << " avg bad " << (used?(avgbad/used) : 0) << std::endl;
		}

		// sort code words
		std::vector<uint64_t> I(n);
		for ( uint64_t i = 0; i < n; ++i )
			I[i] = i;
		DComparator const comp(d,VV);
		std::sort(I.begin(),I.end(),comp);

		// check for collisions
		uint64_t low = 0;
		while ( low < I.size() )
		{
			uint64_t high = low+1;
			while ( high < I.size() && comp.eq(I[low],I[high]) )
				++high;

			if ( high-low > 1 )
			{
				if ( ! collisionfound )
					std::cerr << "[S] collision !" << std::endl;
				collisionfound = true;
			}

			low = high;
		}

		if ( checkHamming )
		{
			// minimum Hamming distance found
			std::vector<uint64_t> M(n,std::numeric_limits<uint64_t>::max());
			// key bit length
			unsigned int const bitlength = d*8;

			// maximum number of errors (Hamming distance) checked for
			unsigned int const maxerr = 7;
			unsigned int const tparts = std::min(maxerr+1,bitlength);
			assert ( tparts <= bitlength );
			unsigned int const tbitsperpart = (bitlength/tparts);
			assert ( tbitsperpart != 0 );
			unsigned int const numparts = (bitlength + tbitsperpart -1)/tbitsperpart;
			unsigned int const bitsperpart = (bitlength + numparts-1)/numparts;

			unsigned int const rest = bitlength % bitsperpart;
			unsigned int const frac = rest ? (bitsperpart-rest) : 0;
			unsigned int const full = numparts - frac;

			for ( unsigned int p = 0; p < numparts; ++p )
			{
				bool const isfull = p < full;
				uint64_t const b_low = isfull ? p*bitsperpart : full*bitsperpart + (p-full)*(bitsperpart-1);
				uint64_t const b_size = isfull ? bitsperpart : (bitsperpart-1);

				DBitComparator const bcomp(d,VV,b_low,b_size);

				std::cerr << "[S] part " << (p+1) << "/" << numparts << " b_low=" << b_low << " b_size=" << b_size << std::endl;

				std::vector<uint64_t> J(n);
				for ( uint64_t i = 0; i < n; ++i )
					J[i] = i;
				std::sort(J.begin(),J.end(),bcomp);

				uint64_t low = 0;
				while ( low < J.size() )
				{
					uint64_t high = low+1;
					while ( high < J.size() && bcomp.get(J[high]) == bcomp.get(J[low]) )
						++high;

					for ( uint64_t i0 = low; i0 < high; ++i0 )
						for ( uint64_t i1 = low; i1 < high; ++i1 )
							if ( i0 != i1 )
							{
								uint64_t const h = bcomp.hamming(J[i0],J[i1]);

								uint64_t const rr = comp.find(J[i0],I); // find rank on I

								// check we found the correct word
								for ( uint64_t zz = 0; zz < comp.d; ++zz )
									assert ( VV [ d * J[i0] + zz ] == VV [ d * I[rr] + zz ]);

								// update minimum distance for code word
								M[rr] = std::min(M[rr],h);
							}

					low = high;
				}
			}

			uint64_t minh = std::numeric_limits<uint64_t>::max();
			for ( uint64_t i = 0; i < M.size(); ++i )
				minh = std::min(minh,M[i]);
			if ( minh <= maxerr )
				std::cerr << "[S] minimum distance " << minh << std::endl;
			else
				std::cerr << "[S] no codeword pair within Hamming distance " << maxerr << std::endl;
		}
	}
}

int bamseqchksumStats(::libmaus2::util::ArgInfo const & arginfo)
{
	if ( isatty(STDIN_FILENO) )
	{
		::libmaus2::exception::LibMausException se;
		se.getStream() << "Refusing to read data from terminal, please redirect standard input to pipe or file." << std::endl;
		se.finish();
		throw se;
	}

	libmaus2::timing::RealTimeClock rtc;
	rtc.start();
	double prevtime = 0;

	int const verbose = arginfo.getValue<int>("verbose",getDefaultVerbose());
	int const checkhamming = arginfo.getValue<int>("checkhamming",getDefaultCheckHamming());
	std::string const hash = arginfo.getValue<std::string>("hash",getDefaultHash());

	// input decoder wrapper
	libmaus2::bambam::BamAlignmentDecoderWrapper::unique_ptr_type decwrapper(
		libmaus2::bambam::BamMultiAlignmentDecoderFactory::construct(arginfo));
	::libmaus2::bambam::BamAlignmentDecoder & dec = decwrapper->getDecoder();
	::libmaus2::bambam::BamHeader const & header = dec.getHeader();

	::libmaus2::bambam::BamAlignment & algn = dec.getAlignment();

	libmaus2::bambam::ChecksumsInterface::unique_ptr_type Pchksums(libmaus2::bambam::ChecksumsFactory::construct(hash,header));
	libmaus2::bambam::ChecksumsInterface & chksums = *Pchksums;

	std::vector<uint8_t> TV;
	std::vector<uint8_t> b_seq_V;

	uint64_t c = 0;
	while ( dec.readAlignment() )
	{
		chksums.update(algn);

		chksums.get_b_seq_all(TV);
		for ( uint64_t i = 0; i < TV.size(); ++i )
			b_seq_V.push_back(TV[i]);

		if ( verbose && (++c & (1024*1024-1)) == 0 )
		{
			double const elapsed = rtc.getElapsedSeconds();
			chksums.printVerbose(std::cerr,c,algn,elapsed-prevtime);
			prevtime = elapsed;

			evaluateStats(TV,b_seq_V,checkhamming);
		}
	}

	chksums.printChecksums(std::cout);

	if ( verbose )
	{
		std::cerr << "[V] run time " << rtc.getElapsedSeconds() << " (" << rtc.formatTime(rtc.getElapsedSeconds()) << ")" << std::endl;
	}

	if ( TV.size() )
	{
		uint64_t const num = b_seq_V.size() / TV.size();
		std::cerr << "[S] num " << num << std::endl;
		evaluateStats(TV,b_seq_V,checkhamming);
	}

	return EXIT_SUCCESS;
}

int bamseqchksum(::libmaus2::util::ArgInfo const & arginfo)
{
	if ( isatty(STDIN_FILENO) )
	{
		::libmaus2::exception::LibMausException se;
		se.getStream() << "Refusing to read data from terminal, please redirect standard input to pipe or file." << std::endl;
		se.finish();
		throw se;
	}

	libmaus2::timing::RealTimeClock rtc;
	rtc.start();
	double prevtime = 0;

	int const verbose = arginfo.getValue<int>("verbose",getDefaultVerbose());
	std::string const hash = arginfo.getValue<std::string>("hash",getDefaultHash());

	// input decoder wrapper
	libmaus2::bambam::BamAlignmentDecoderWrapper::unique_ptr_type decwrapper(
		libmaus2::bambam::BamMultiAlignmentDecoderFactory::construct(arginfo));
	::libmaus2::bambam::BamAlignmentDecoder & dec = decwrapper->getDecoder();
	::libmaus2::bambam::BamHeader const & header = dec.getHeader();

	::libmaus2::bambam::BamAlignment & algn = dec.getAlignment();

	libmaus2::bambam::ChecksumsInterface::unique_ptr_type Pchksums(libmaus2::bambam::ChecksumsFactory::construct(hash,header));
	libmaus2::bambam::ChecksumsInterface & chksums = *Pchksums;

	uint64_t c = 0;
	while ( dec.readAlignment() )
	{
		chksums.update(algn);

		if ( verbose && (++c & (1024*1024-1)) == 0 )
		{
			double const elapsed = rtc.getElapsedSeconds();
			chksums.printVerbose(std::cerr,c,algn,elapsed-prevtime);
			prevtime = elapsed;
		}
	}

	chksums.printChecksums(std::cout);

	if ( verbose )
	{
		std::cerr << "[V] run time " << rtc.getElapsedSeconds() << " (" << rtc.formatTime(rtc.getElapsedSeconds()) << ")" << std::endl;
	}

	return EXIT_SUCCESS;
}

#if defined(BIOBAMBAM_HAVE_GMP)
template<size_t k>
static libmaus2::math::UnsignedInteger<k> convertNumber(mpz_t const & gmpnum)
{
	size_t const numbitsperel = 8 * sizeof(uint32_t);
	size_t const numwords = (mpz_sizeinbase(gmpnum,2) + numbitsperel - 1) / numbitsperel;
	libmaus2::autoarray::AutoArray<uint32_t> A(numwords,false);
	size_t countp = numwords;
	mpz_export(A.begin(),&countp,-1,sizeof(uint32_t),0,0,gmpnum);
	libmaus2::math::UnsignedInteger<k> U;
	for ( size_t i = 0; i < std::min(countp,static_cast<size_t>(k)); ++i )
		U[i] = A[i];
	return U;
}
// search for next prime number larger than 2^(32*k) using probabilistic algorithm in gmp library
template<size_t k>
libmaus2::math::UnsignedInteger<k+1> nextPrime()
{
	libmaus2::math::UnsignedInteger<k+1> U(1);
	U <<= (k*32);

	mpz_t gmpU;
	mpz_init(gmpU);
	mpz_import(gmpU,k+1,-1 /* least sign first */,sizeof(uint32_t),0 /* native endianess */,0 /* don't skip bits */, U.getWords());

	while ( mpz_probab_prime_p(gmpU,200) == 0 )
		mpz_add_ui(gmpU,gmpU,1);

	libmaus2::math::UnsignedInteger<k+1> const R = convertNumber<k+1>(gmpU);

	mpz_clear(gmpU);

	return R;
}
#endif

int main(int argc, char * argv[])
{
	try
	{
		::libmaus2::util::ArgInfo const arginfo(argc,argv);

		for ( uint64_t i = 0; i < arginfo.restargs.size(); ++i )
			if (
				arginfo.restargs[i] == "-v"
				||
				arginfo.restargs[i] == "--version"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				return EXIT_SUCCESS;
			}
			else if (
				arginfo.restargs[i] == "-h"
				||
				arginfo.restargs[i] == "--help"
			)
			{
				std::cerr << ::biobambam2::Licensing::license();
				std::cerr << std::endl;
				std::cerr << "Key=Value pairs:" << std::endl;
				std::cerr << std::endl;

				std::vector< std::pair<std::string,std::string> > V;

				V.push_back ( std::pair<std::string,std::string> ( "verbose=<["+::biobambam2::Licensing::formatNumber(getDefaultVerbose())+"]>", "print progress information" ) );
				#if defined(BIOBAMBAM_LIBMAUS2_HAVE_IO_LIB)
				V.push_back ( std::pair<std::string,std::string> ( std::string("inputformat=<[")+getDefaultInputFormat()+"]>", "input format: cram, bam or sam" ) );
				V.push_back ( std::pair<std::string,std::string> ( "reference=<[]>", "name of reference FastA in case of inputformat=cram" ) );
				#else
				V.push_back ( std::pair<std::string,std::string> ( "inputformat=<[bam]>", "input format: bam" ) );
				#endif

				V.push_back ( std::pair<std::string,std::string> ( std::string("hash=<[")+getDefaultHash()+"]>", "hash digest function: " + libmaus2::bambam::ChecksumsFactory::getSupportedHashVariantsList()) );

				::biobambam2::Licensing::printMap(std::cerr,V);

				std::cerr << std::endl;

				return EXIT_SUCCESS;
			}

		if ( arginfo.getValue<int>("stats",getDefaultStats()) )
			return bamseqchksumStats(arginfo);
		else
			return bamseqchksum(arginfo);
	}
	catch(std::exception const & ex)
	{
		std::cerr << ex.what() << std::endl;
		return EXIT_FAILURE;
	}
}
